Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp

```diff
@@ -5627,6 +5628,77 @@ void WebGLRenderingContextBase::forceContextLost()
     forceLostContext(WebGLRenderingContextBase::RealLostContext);
 }
 
+static ASCIILiteral debugMessageTypeToString(GCGLenum type)
+{
+    switch (type) {
+    case GraphicsContextGL::DEBUG_TYPE_ERROR:
+        return "error"_s;
+    case GraphicsContextGL::DEBUG_TYPE_DEPRECATED_BEHAVIOR:
+        return "deprecated behavior"_s;
+    case GraphicsContextGL::DEBUG_TYPE_UNDEFINED_BEHAVIOR:
+        return "undefined behavior"_s;
+    case GraphicsContextGL::DEBUG_TYPE_PORTABILITY:
+        return "portability"_s;
+    case GraphicsContextGL::DEBUG_TYPE_PERFORMANCE:
+        return "performance"_s;
+    case GraphicsContextGL::DEBUG_TYPE_MARKER:
+        return "marker"_s;
+    case GraphicsContextGL::DEBUG_TYPE_OTHER:
+        return "other"_s;
+    default:
+        ASSERT_NOT_REACHED();
+        return "unknown"_s;
+    }
+}
+
+static ASCIILiteral debugIdToString(GCGLenum id)
+{
+    switch (id) {
+    case GraphicsContextGL::INVALID_ENUM:
+        return "INVALID_ENUM"_s;
+    case GraphicsContextGL::INVALID_VALUE:
+        return "INVALID_VALUE"_s;
+    case GraphicsContextGL::INVALID_OPERATION:
+        return "INVALID_OPERATION"_s;
+    case GraphicsContextGL::OUT_OF_MEMORY:
+        return "OUT_OF_MEMORY"_s;
+    case GraphicsContextGL::INVALID_FRAMEBUFFER_OPERATION:
+        return "INVALID_FRAMEBUFFER_OPERATION"_s;
+    case GraphicsContextGL::CONTEXT_LOST_WEBGL:
+        return "CONTEXT_LOST_WEBGL"_s;
+    default:
+        return "unknown"_s;
+    }
+}
+
+static ASCIILiteral debugMessageSeverityToString(GCGLenum severity)
+{
+    switch (severity) {
+    case GraphicsContextGL::DEBUG_SEVERITY_HIGH:
+        return "high"_s;
+    case GraphicsContextGL::DEBUG_SEVERITY_MEDIUM:
+        return "medium"_s;
+    case GraphicsContextGL::DEBUG_SEVERITY_LOW:
+        return "low"_s;
+    case GraphicsContextGL::DEBUG_SEVERITY_NOTIFICATION:
+        return "notification"_s;
+    default:
+        ASSERT_NOT_REACHED();
+        return "unknown"_s;
+    }
+}
+
+void WebGLRenderingContextBase::addDebugMessage(GCGLenum type, GCGLenum id, GCGLenum severity, const String& message)
+{
+    if (!shouldPrintToConsole())
+        return;
+
+    if (type == GraphicsContextGL::DEBUG_TYPE_ERROR)
+        printToConsole(MessageLevel::Error, makeString("WebGL: "_s, debugIdToString(id), ": "_s, message));
```

Comment:
MessageLevel::Error will produce a JS stack shot which we cannot do, as this is called within the IPC message dispatch.
You need to inline enough stuff from printToConsole to somewhat match the format that it's natural for the reader.